adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+            /* Codes_SRS_TLSIO_APPLEIOS_32_017: [ The tlsio_appleios_create shall receive the connection configuration (TLSIO_CONFIG). ]*/
+            TLSIO_CONFIG* tlsio_config = (TLSIO_CONFIG*)io_create_parameters;
+            
+            if (tlsio_config->hostname == NULL)
+            {
+                LogError("Host name was not provided");
+            }
+            else {
+                /* Codes_SRS_TLSIO_APPLEIOS_32_016: [ The tlsio_appleios_create shall initialize all callback pointers as NULL. ]*/
+                tlsio_instance->on_io_open_complete = NULL;
+                tlsio_instance->on_io_open_complete_context = NULL;
+                tlsio_instance->on_bytes_received = NULL;
+                tlsio_instance->on_bytes_received_context = NULL;
+                tlsio_instance->on_io_error = NULL;
+                tlsio_instance->on_io_error_context = NULL;
+                tlsio_instance->on_io_close_complete = NULL;
+                tlsio_instance->on_io_close_complete_context = NULL;
+
+                /* Codes_SRS_TLSIO_APPLEIOS_32_020: [ If tlsio_appleios_create get success to create the tlsio instance, it shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSED. ]*/
+                tlsio_instance->state = TLSIO_APPLEIOS_STATE_CLOSED;
+
+                tlsio_instance->port = (uint16_t)tlsio_config->port;
+                tlsio_instance->sockRead = NULL;
+                tlsio_instance->sockWrite = NULL;
+
+                /* Codes_SRS_TLSIO_APPLEIOS_32_018: [ The tlsio_appleios_create shall store provided hostName. ]*/
+                tlsio_instance->remote_host = CFStringCreateWithCString(NULL, tlsio_config->hostname, kCFStringEncodingUTF8);
+                
+                if (tlsio_instance->remote_host == NULL)
+                {
+                    LogError("Unable to allocate string for host name");
+                    free(tlsio_instance);
+                    tlsio_instance = NULL;
+                }
+            }
+        }
+    }
+
+    /* Codes_SRS_TLSIO_APPLEIOS_32_010: [ The tlsio_appleios_create shall return a non-NULL handle on success. ]*/
+    return (CONCRETE_IO_HANDLE)tlsio_instance;
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_021: [ The tlsio_appleios_destroy shall destroy a created instance of the tlsio for iOS identified by the CONCRETE_IO_HANDLE. ]*/
+void tlsio_appleios_destroy(CONCRETE_IO_HANDLE tlsio_handle)
+{
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if (tlsio_handle == NULL)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_024: [ If the tlsio_handle is NULL, the tlsio_appleios_destroy shall not do anything. ]*/
+        LogError("NULL TLS handle.");
+    }
+    else
+    {
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPEN) ||
+            (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
```

Comment:
Please check if you don't need to force close the connection on these status... And for TLSIO_APPLEIOS_STATE_ERROR too.